starlark/src/values/string.rs

```diff
@@ -17,8 +17,252 @@ use crate::values::*;
+                // %(name)s or %s
+                enum NamedOrPositional {
+                    Named(String),
+                    Positional,
+                }
+                let (named_or_positional, format_char) = if next == '(' {
+                    let mut name = String::new();
+                    loop {
+                        match chars.next() {
+                            None => return Err(ValueError::InterpolationFormat),
+                            Some(')') => {
+                                break;
+                            }
+                            Some(c) => name.push(c),
+                        }
+                    }
+                    match chars.next() {
+                        Some(c) => (NamedOrPositional::Named(name), c),
+                        None => return Err(ValueError::InterpolationFormat),
+                    }
+                } else {
+                    (NamedOrPositional::Positional, next)
+                };
+                let format = match format_char {
+                    's' => ArgFormat::Str,
+                    'r' => ArgFormat::Repr,
+                    'd' | 'i' => ArgFormat::Dec,
+                    'o' => ArgFormat::Octal,
+                    'x' => ArgFormat::HexLower,
+                    'X' => ArgFormat::HexUpper,
+                    'c' => ArgFormat::Char,
+                    '%' => match named_or_positional {
+                        NamedOrPositional::Positional => {
+                            result.append_literal('%');
+                            continue;
+                        }
+                        NamedOrPositional::Named(_) => {
+                            return Err(ValueError::InterpolationFormat);
+                        }
+                    },
+                    _ => return Err(ValueError::InterpolationFormat),
+                };
+                result = match (named_or_positional, result) {
+                    (NamedOrPositional::Named(name), ArgsFormat::Named(init, mut args)) => {
+                        args.push((name, format, String::new()));
+                        ArgsFormat::Named(init, args)
+                    }
+                    (NamedOrPositional::Positional, ArgsFormat::Positional(init, mut args)) => {
+                        args.push((format, String::new()));
+                        ArgsFormat::Positional(init, args)
+                    }
+                    (NamedOrPositional::Named(name), ArgsFormat::Positional(init, args)) => {
+                        if args.is_empty() {
+                            ArgsFormat::Named(init, vec![(name, format, String::new())])
+                        } else {
+                            return Err(ValueError::InterpolationFormat);
+                        }
+                    }
+                    (NamedOrPositional::Positional, ArgsFormat::Named(init, args)) => {
+                        if args.is_empty() {
+                            // unreachable
```

Comment:
Since it is unreachable, use the `unreachable!()` macro. In that case I would also omit the if/else.

Reply:
The code was rewritten differently. Named and positional are no longer exclusive to each other.